### ISSUE TITLE
HMM with terminal state

### ISSUE THREAD

**Issue by alinafere** [luisdamiano/gsoc17-hhmm#2]
Hi Luis, 
I am trying to code an HMM where certain states are absorbing states, along the lines of your HHMM model. Here is a summary of a simplified version of the model:
1. There are 50 individuals, each with T=15 observations.
2. The observation variable y is binary (success/ failure). 
3. There are K=2 states, and state 2 is a terminal state.  

I am attaching an R file, with both simulation and estimation codes for this model. I get very strange results, and a reversal in the start, emission and transition probabilities between state 1 and state 2. 
Would really appreciate your help to figure out what's going on. 
Cheers, 
Alina

```{r setup, include=FALSE}
rm(list=ls())
library(rstan)
library(dplyr)
library(magrittr)
library(ggplot2)

rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())

## hmm terminal state  no covariates ------
hmm_terminal=" 
functions {
vector normalize(vector x) {
return x / sum(x);
}
}

data {
int<lower=1> N;                   // number of observations
int<lower=1> IND;                  // number of indivivuals
int<lower=1> K;                   // number of hidden states                  // number of predictors in the choice eq.
int<lower=1> D;                  // number of possible choices
int<lower=1> T[N];               // vector for the time series per individual
int<lower=0, upper=1> lastT[N];  // vector for the time series per individual
int<lower=1> ID[N];              // vector of individual index
int<lower=0, upper=1> y[N];      // observations for the choice eq.
}


parameters {
simplex[K] Aij[K];
simplex[K] pi0;
simplex[D] phi_k[K];
}

transformed parameters {
vector[K] unalpha_tk[N];
vector[K] unalpha_Tkn[IND];

for (n in 1:N)
{ // Forwards algorithm log p(z_t = j | x_{1:t})
real accumulator[K];

// at T=1    
if(T[n]==1)
{
  for (j in 1:K) {
  unalpha_tk[n][j] = log(pi0[j]) + log(phi_k[j, y[n]+1]);
  }
}else{
  for (j in 1:K) { // j = current (t)
  for (i in 1:K) { // i = previous (t-1)

// if at terminnal state K, transition to other states is 0, the likelihood is not augmented;
 if (i==K && j!=K) 
  accumulator[i] = unalpha_tk[n-1, i];
// the likelihood of staying in state K is 1, thus we only augment with the emission prob;
  else if (i==K && j==K)
  accumulator[i] = unalpha_tk[n-1, i]+ log(phi_k[j, y[n]+1]);
// if not at terminal state, add transition and emission probs;
  else accumulator[i] = unalpha_tk[n-1, i] + log(Aij[i,j]) + log(phi_k[j, y[n]+1]);
  }
  unalpha_tk[n, j] = log_sum_exp(accumulator);
  }

}
if(lastT[n]==1) 
{unalpha_Tkn[ID[n]]=unalpha_tk[n];}

} // Forward
}

model {
for (i in 1:IND)
target += log_sum_exp(unalpha_Tkn[i]); // Note: update based only on last unalpha_tk

}"

m = stan_model(model_code = hmm_terminal)

### SIMULATE DATA ----
### fullsim data
set.seed(9000)
NID=50
ID=1:NID
T.sim = rep(15, NID) #rgeom(NID, 0.25)+1
N=sum(T.sim)
IND=NULL
for (i in 1:NID)
{ind=rep(ID[i], T.sim[i])
IND=c(IND, ind)}

## number of states
K= 2
## start probability
pi0 = c(0.8, 0.2)
# transition matrix
Aij <- matrix(c(.6, .4, 0,1), nrow = K, ncol = K, byrow=T)
## simulate states
ztrue=NULL
probs=NULL
for (i in 1:NID){
  zind=rep(0, T.sim[i])
  zind[1]=sample(x = 1:K, size = 1, prob = pi0)
  if(T.sim[i]>1){
    for (t in 2:T.sim[i]) {
       zind[t] <- sample(x = 1:K, size = 1, replace = F, prob = Aij[zind[t-1], ])
    }
  }
  print(table(zind))
  ztrue=c(ztrue, zind)
  probs=rbind(probs, Aij)
}

## simulate y per state
y <- vector("numeric", N)
emprob=c(.2, .5)
phi_k=NULL
for (t in 1:N) {
  # emission probs - prob to apply given state
  phi_tk=emprob[ztrue[t]]
  y[t] <- rbinom(1, size = 1, prob = phi_tk)
  phi_k=c(phi_k,phi_tk)
}
sum(y)
d=data_frame(state=as.factor(ztrue), apply=as.factor(y), phi_k) 

d %>% 
  ggplot(aes(x = phi_k, fill =applynow)) +
  geom_histogram(alpha = 0.5, position="dodge", binwidth = 0.1) + theme_bw() +
  facet_wrap(~state)+
  labs(title = "Probability of applying when state=1 vs. state=2", x = "p(apply=1)", y = "Count")

table(ztrue, y)

hmm_simdata=data.frame(IND=IND)
hmm_simdata %<>% group_by(IND) %>% mutate(T.length=row_number(), lastT=ifelse(row_number()==n(), 1,0)) %>% ungroup()
head(hmm_simdata)
hmm_simdata %<>% 
  mutate(group_id = group_indices(., IND), Obs=1:n())


# Stan Data -------------------------------------------------------------------
ID=pull(hmm_simdata, group_id)
T.length=pull(hmm_simdata, T.length)
lastT=pull(hmm_simdata, lastT)

stan.data = list(
  N=N,                   # number of observations
  IND=length(unique(IND)),             # number of individuals
  K=2,                               # number of hidden states      
  D=2,                               # number of possible choices
  T=T.length,                       # vector for the time series per individual
  ID=ID,                            # vector of individual index
  y = y,                   # observations for the choice eq.
  lastT=lastT
  )

### Run the model ------

hmm2S_terminalSt <- stan(model_code = hmm_terminal, data = stan.data,
                       verbose = TRUE, chains = 2,
                       seed = 9000, iter = 1500, warmup =1200, control = list(adapt_delta=0.99, max_treedepth=15))

stfit=hmm2S_terminalSt
phis=summary(stfit, pars = c("lp__","pi0", "phi_k", "Aij"), probs=c(0.025, 0.975))$summary
phis
Qs=summary(stfit, pars = c("Aij", "lp__"), probs=c(0.025, 0.975))$summary
Qs

traceplot(stfit, pars = c("lp__", "pi0", "phi_k"), inc_warmup = FALSE, ncol=4)
traceplot(stfit, pars = c("Aij","lp__"), inc_warmup = F, ncol=4)

```

**Comment by michaelweylandt** [comment]
Hi Alina, 

I haven't run your code, but a few things jump out at me: 

1) If you only have two states, one of which is known to be absorbing, it might be easier to code your model as some sort of change-point model rather than a full HMM. 
2) I don't think your approach of ignoring transitions out of state `K` in the forward algorithm will work as written. There's still a simplex constraint on the transition probabilities and Stan will try to put some probability on transition out of that state. Since that won't get reflected in the likelihood, you've basically got an identifiability issue, which will confuse the heck out of HMC. 

A better approach would be to make a `K-1` by `K` matrix of transition probabilities (each row a simplex) and then use the `transformed parameters` block to add an additional row of all zeros to get the full transition matrix. Does that make sense?

**Comment by alinafere** [comment]
Hi Michael, 
Thanks a lot for your suggestions. 
I tried 2. at it worked beautifully. Now everything is well identified. 
I will think about 1. as well, seems more reasonable for this 2 states model. The model I posted is a simplified version of what I have, which is a an input-output HMM, and I do plan to try it with 3 states. I'll see which model works better for this data. 
Thanks again.

Happy New Year!
